Small performance optimizations

Torathion opened the pull request:
Hi, I wanted to contribute to this amazing project and came across some tiny performance optimizations that could be made. I tried not to change too much code or break your coding style. Those are very minor changes that only benefit projects and tools processing a large number of paths and globs. 

The PR mainly focuses on stabilizing the performance by:
- inlining unnecessary variables
- reducing overhead by reducing the number of object property calls like `.length` or `.root`
- avoiding accidental duplicate calculations like the windows platform check or the loop length of the pattern normalization
- changing the fdir path transformation from creating a new array via `map` to transforming it in-place

Also, tinyglobby is now ~25 bytes smaller.

Torathion commented:
Sorry for the extra linting commits. Biome doesn't work on my machine.

SuperchupuDev commented:
interesting, i do value performance over code style, but i'd like to first check if this makes any meaningful perf improvement, the vast majority of the glob time in general is spent doing I/O i think which is the area that could benefit the most from optimizations. i'll check this out in more detail later, there's some work i need to do first. thanks for the effort!

Torathion commented:
I tested it with your test fixture and it roughly was the same performance, but my version was more stable in terms of performance, so I do believe you see the greater benefit if you create a fixture with a ton of files and glob patterns, because I only adjusted the performance inside loops.

I'm also still trying to figure out an alternative to `posix.relative` as you mentioned the performance issues inside your code. But since `path.relative` and `path.posix.relative` have a different behavior, I wait until you fixed the windows parsing issues for me to experiment.

Torathion commented:
I've updated my performance update to include the changes from #76 .

There are only two major things I've changed:
- Inlined a variable and renamed `properties` to `props` to reduce the upload and download package size (because you don't minify it)
- Optimized the `getPartialMatcher` method to cache splitted patterns, replacing the `map` comfort function with pre-allocated arrays and reduced the number of unnecessary object property accesses.

Over all, it should heavily reduce the GC work time and memory overhead, depending an how many inputs are processed through partial matching.

SuperchupuDev commented:
thank you!! it seems like your branch wasn't properly rebased though, like it somehow doesn't detect the new commits from main as not part of this pr, which makes reviewing harder (check the files tab on github)

SuperchupuDev commented on src/index.ts:
```diff
@@ -241,19 +248,15 @@ function crawl(options: GlobOptions, cwd: string, sync: boolean) {
-  properties.root = properties.root.replace(BACKSLASHES, '');
-  const api = new fdir(fdirOptions).crawl(properties.root);
+  props.root = props.root.replace(BACKSLASHES, '');
+  const root = props.root;
```
i do not think adding new variables that refer to object properties would really help performance, unless reading properties in javascript are somehow not performant. if you can show me in any way that creating a new variable that copies an object property is faster than just accessing the property itself i'll allow it

Torathion replied:
This has to benefits:

1. Yes, property calls are dynamic, because objects are dynamically stored objects pointing to different memory cells. Consider this test:

```js
const array = [0, 1, 2, 3, 4, 5, 6, 7]

console.time('property call')
for (let i = 0; i < 10000; i++) array.length
console.timeEnd('property call') // 0.176ms
console.time('variable')
const length = array.length
for (let i = 0; i < 10000; i++) {
    length
}
console.timeEnd('variable') // 0.066ms
```

2. Property names can not be minified.

```js

longFunctionName(longObjectName.longPropertyName) // will minify to f(e.longPropertyName)

const longVariableName = longObjectName.longPropertyName
longFunctionName(longVariableName) // will minify to f(e)
```

The second one is only useful if you use the specific property multiple times.

Torathion replied:
Also, I wanted to write `const root = props.root = props.root.replace(BACKSLASHES, '');`, but your biome config wouldn't let me.

SuperchupuDev commented on src/utils.ts:
```diff
@@ -48,7 +49,7 @@ export function getPartialMatcher(patterns: string[], options?: PartialMatcherOp
         j++;
       }
-      if (j === inputParts.length) {
+      if (j === inputPatternCount) {
```
same here

Torathion replied:
Basic loop optimizations. Read [here](https://www.linkedin.com/pulse/optimizing-performance-best-practices-loops-javascript-bowin-p) for more.

Torathion replied:
Now that I come to think of it: I could inline the array length variable in the for loop head.

```js
for (let i = 0, len = array.length; i < len; i++)
```

I've tested it in my projects and I have saved a couple hundred of bytes by inlining it.

SuperchupuDev commented on src/utils.ts:
```diff
@@ -59,12 +60,23 @@ export function getPartialMatcher(patterns: string[], options?: PartialMatcherOp
 // #endregion
 
 // #region splitPattern
+// Make the options a global constant to reduce GC work
+const SplitPatternOptions = { parts: true }
```
sounds redundant considering objects are used in multiple places and splitting patterns aren't done in performance critical areas. also should follow camelCase and not PascalCase

SuperchupuDev commented on src/utils.ts:
```diff
@@ -59,12 +60,23 @@ export function getPartialMatcher(patterns: string[], options?: PartialMatcherOp
+const SplitPatternOptions = { parts: true }
+
 // if a pattern has no slashes outside glob symbols, results.parts is []
+/**
+ * Splits a path pattern to its parts.
+ * If the given path has no slashes outside glob symbols, the given path will be interpreted as the only part.
+ *
+ * @param path - target path to split.
+ * @returns the parts of the path.
+ */
```
i think these comments are irrelevant, i believe the function is straightforward enough to be understood easily. could consider it if it was a function exposed to users but it's completely internal

Torathion replied:
Those are standard [tsdoc](https://tsdoc.org/) documentation comments. You can change the comments to be more precise, because I have no idea what you mean with patterns.

Those documentation comments are necessary to produce useful IDE tooltips
![image](https://github.com/user-attachments/assets/dbcc8356-2f77-4d2f-b50e-69ba89a772b9)

SuperchupuDev commented on src/utils.ts:
```diff
@@ -119,3 +132,13 @@ export function isDynamicPattern(pattern: string, options?: { caseSensitiveMatch
 }
 // #endregion
+// #region log
+/**
+ *  Logs the current task.
+ *
+ *  @param task - the current task.
+ */
```
comment is also irrelevant

Torathion replied:
Again, standard documentation comments for a cleaner project.

SuperchupuDev replied:
still, an internal function called `log` is pretty self explanatory, there is nothing to be made clearer here

Torathion replied:
I think it's the best of both worlds if you decide how to design the comments, maybe even with the tsdoc consideration I suggested.

SuperchupuDev reviewed (commented):
overall looking good, but there are some things that might virtually not change performance and would increase lines for (imo) not a good enough reason

SuperchupuDev commented:
wow, i sure have learned something today about how JS optimizes stuff. thanks a lot for diving this deep into javascript microoptimizations! javascript is such an odd language internally :-). i'll remove the internal tsdoc comments for now but i can add them back if we find a need for internal documentation. the codebase is so (relatively) small right now i don't think they are needed

Torathion commented:
Hey, another thing: You've changed how the partial matcher gets his regexes. It should actually look like this:

```ts
  const patternCount = patterns.length
  // Cache splitted patterns to reuse them when parsing input paths
  const splittedPatterns: string[][] = new Array(patternCount)
  const regexes: RegExp[][] = new Array(patternCount)

  for (let i = 0; i < patternCount; i++) {
    const pattern = patterns[i]
    const splittedPattern = splitPattern(pattern)
    splittedPatterns[i] = splittedPattern
    const partCount = splittedPattern.length
    const partRegexes: RegExp[] = new Array(partCount)

    for (let j = 0; j < partCount; j++) {
      partRegexes[j] = picomatch.makeRe(splittedPattern[j], options)
    }

    regexes[i] = partRegexes
  }
```

I did this out of two reasons:

- Array comfort functions like `map` are slow due to frequent function calls "penetrating" the call stack. This creates extra overhead. Native loops still are the fastest way to iterate through stuff.
- I used memory pre-allocated arrays with `new Array(length)`. `[]` is the equivalent of `new Array(0)`. This saves frequent memory rewrites due to array size changes. So `push` is slower than using indices.

If you are concerned about the final bundle size, I would always check how much bigger it is when minifying it.

On bundlephobia, [tinyglobby](https://bundlephobia.com/package/tinyglobby@0.2.10) is still half as big as [fast-glob](https://bundlephobia.com/package/fast-glob@3.3.3), so performance should still be more prioritized. My loops only added like 500 bytes or so.

SuperchupuDev commented:
ah, that part was changed when rebasing. didn't the linkedin post you sent mention that array methods like `.map` are faster than loops? https://www.linkedin.com/pulse/optimizing-performance-best-practices-loops-javascript-bowin-p

regarding `Array(length)`, i'm not really sure if preallocation really works in javascript in terms of better performance. someone from e18e told me that they are pretty sure it doesn't

Torathion commented:
I can tell you that the linkedin post is mostly wrong about this. While comfort functions like `sort` and `find` and `filter` are indeed faster, `forEach` and `map` are slower.

`forEach` test:
```js

// Create a 9 x 9 matrix
const array = Array.from({ length: 9 }, () => Array.from({ length: 9 }, () => Math.random()))

console.time('native')
let sum1 = 0
for (let i = 0; i < 9; i++) {
    for (let j = 0; j < 9; j++) {
        sum1 += array[i][j]
    }
}
console.timeEnd('native') // 0.012ms

console.time('comfort functions')
let sum2 = 0
array.forEach(arr => arr.forEach(value => sum2 += value))
console.timeEnd('comfort functions') // 0.032ms
```

`map` test:

```js

// Create a 9 x 9 matrix
const array = Array.from({ length: 9 }, () => Array.from({ length: 9 }, () => Math.random()))

console.time('native')
const newMatrix1 = []
for (let i = 0; i < 9; i++) {
    const row = []
    for (let j = 0; j < 9; j++) {
        row.push(array[i][j])
    }
    newMatrix1.push(row)
}
console.timeEnd('native') // 0.02ms

console.time('comfort functions')
const newMatrix2 = array.map(arr => arr.map(value => value))
console.timeEnd('comfort functions') // 0.03ms
```

This is due to array comfort functions entering and leaving a function on EVERY item. This creates overhead due to using the call stack. Those functions also have more chances to trigger the GC as we leave more functions (blocks). For example, on time the GC was triggered inside the comfort functions test and it doubled the time from around 0.030ms to 0.065ms.

For the pre-allocation / push thing, I had to make sure node.js doesn't screw with me due to preemptive optimizations, but yes, pre-allocations are faster.

```js
// Create a 81 x 81 matrix
const array = Array.from({ length: 81 }, () => Array.from({ length: 81 }, () => Math.random()))

console.time('push')
const newMatrix2 = []
for (let i = 0; i < 81; i++) {
    const row = []
    for (let j = 0; j < 81; j++) {
        row.push(array[i][j])
    }
    newMatrix2.push(row)
}
console.timeEnd('push') // 0.45ms
console.time('allocated')
const newMatrix1 = new Array(81)
for (let i = 0; i < 81; i++) {
    const row = new Array(81)
    for (let j = 0; j < 81; j++) {
        row[j] = array[i][j]
    }
    newMatrix1[i] = row
}
console.timeEnd('allocated') // 0.295ms
```

Funnily enough, if I switch the cases in my test file, they are equally as fast. Maybe that's why the person from e18e wasn't sure. In my previous tests, I've gathered that on extremely large arrays with thousands of items, it doesn't matter anymore as there are less memory re-writes. I believe the memory rewrites that are due to array size changes use an internal buffer size for the array. And if this buffer size is met, a memory re-write occurs with the buffer doubling in size or so. So pre-allocation is way more effective on smaller arrays.

SuperchupuDev commented:
according to https://dev.to/henryjw/populating-a-pre-allocated-array-slower-than-a-pushing-to-a-regular-array-4l43 array pushes are more optimized in v8 than direct assignments, but it's a 5 year old post so who knows if it's still the case. we could try to measure the code samples given there

Torathion commented:
It's still faster. I copied the code and got a stackoverflow error, so I changed the number of elements to `10**7`

```js
const NUMBER_OF_ELEMENTS = 10**7
const ELEMENT_LEN_BYTES = 4

const array = []

console.time('array')

for (let i = 1; i <= NUMBER_OF_ELEMENTS; i++) {
    array.push(i)
}

console.timeEnd('array')

const preAllocatedArray = new Array(NUMBER_OF_ELEMENTS)

console.time('pre-allocated array')

for (let i = 1; i <= NUMBER_OF_ELEMENTS; i++) {
    preAllocatedArray[i - 1] = i
}

console.timeEnd('pre-allocated array')

const intArray = new Uint32Array(NUMBER_OF_ELEMENTS)

console.time('int array')

for (let i = 0; i < NUMBER_OF_ELEMENTS; i++) {
    intArray[i] = i + 1
}

console.timeEnd('int array')


const buffer = Buffer.alloc(NUMBER_OF_ELEMENTS * ELEMENT_LEN_BYTES, 0)

console.time('buffer')

for (let i = 1, offset = 0; i <= NUMBER_OF_ELEMENTS; i++) {
    offset = buffer.writeUInt32BE(i, offset)
}

console.timeEnd('buffer')
```

That's what I got:

```powershell
array: 148.674ms
pre-allocated array: 11.094ms
int array: 13.012ms
buffer: 73.315ms
```

SuperchupuDev commented:
interesting, me too

Torathion commented:
I guess it has to do with the actual number. Maybe it was much slower with 10*8, because it surpassed some kind of number barrier for an internal data type and they tried to handle it somehow back then.

SuperchupuDev commented:
might be, changing it to 10**8 doesn't crash for me and makes the preallocated one much slower
![image](https://github.com/user-attachments/assets/bbebb7d2-54a0-4b0f-8628-303032f0042d)

Torathion commented:
Which only confirms that pre-allocated arrays are faster on smaller arrays. 

If there is nothing else to do, I'd be happy if this could be merged 🙏

SuperchupuDev reviewed (approved):
asking around e18e some more it doesn't look like there is a definitive answer, but many have said preallocated is probably faster. will keep investigating but this pr should be good to go. thanks a lot for your work and patience! i think i've learned a bit more about js internals with this :P
